src/main/java/appeng/thirdparty/codechicken/lib/model/Quad.java

```diff
@@ -0,0 +1,557 @@
+
+		if( format.hasNormal )
+		{
+			for( Vertex vertex : vertices )
+			{
+				vertex.normal[0] = normal.x;
+				vertex.normal[1] = normal.y;
+				vertex.normal[2] = normal.z;
+				vertex.normal[3] = 0;
+			}
+		}
+		orientation = EnumFacing.getFacingFromVector( normal.x, normal.y, normal.z );
+	}
+
+	/**
+	 * Used to create a new quad complete copy of this one.
+	 *
+	 * @return The new quad.
+	 */
+	public Quad copy()
+	{
+		if( !full )
+		{
+			throw new RuntimeException( "Only copying full quads is supported." );
+		}
+		Quad quad = new Quad( format );
+		quad.tintIndex = tintIndex;
+		quad.orientation = orientation;
+		quad.diffuseLighting = diffuseLighting;
+		quad.sprite = sprite;
+		quad.full = true;
+		for( int i = 0; i < 4; i++ )
+		{
+			quad.vertices[i] = vertices[i].copy();
+		}
+		return quad;
+	}
+
+	/**
+	 * Copies the data inside the given quad to this one. This ignores VertexFormat, please make sure your quads are in the same format.
+	 *
+	 * @param quad The Quad to copy from.
+	 *
+	 * @return This quad.
+	 */
+	public Quad copyFrom( Quad quad )
+	{
+		tintIndex = quad.tintIndex;
+		orientation = quad.orientation;
+		diffuseLighting = quad.diffuseLighting;
+		sprite = quad.sprite;
+		full = quad.full;
+		for( int v = 0; v < 4; v++ )
+		{
+			for( int e = 0; e < format.elementCount; e++ )
+			{
+				System.arraycopy( quad.vertices[v].raw[e], 0, vertices[v].raw[e], 0, 4 );
+			}
+		}
+		return this;
+	}
+
+	/**
+	 * Reset the quad to the new format.
+	 *
+	 * @param format The new format.
+	 */
+	public void reset( CachedFormat format )
+	{
+		this.format = format;
+		tintIndex = -1;
+		orientation = null;
+		diffuseLighting = true;
+		sprite = null;
+		for( int i = 0; i < vertices.length; i++ )
+		{
+			Vertex v = vertices[i];
+			if( v == null )
+			{
+				vertices[i] = v = new Vertex( format );
+			}
+			v.reset( format );
+		}
+		vertexIndex = 0;
+		full = false;
+	}
+
+	/**
+	 * Bakes this Quad to a BakedQuad.
+	 *
+	 * @return The BakedQuad.
+	 */
+	public BakedQuad bake()
+	{
+		int[] packedData = new int[format.format.getNextOffset()];
+		for( int v = 0; v < 4; v++ )
+		{
+			for( int e = 0; e < format.elementCount; e++ )
+			{
+				LightUtil.pack( vertices[v].raw[e], packedData, format.format, v, e );
+			}
+		}
+		return new BakedQuad( packedData, tintIndex, orientation, sprite, diffuseLighting, format.format );
+	}
+
+	/**
+	 * Bakes this quad to an UnpackedBakedQuad.
+	 *
+	 * @return The UnpackedBakedQuad.
+	 */
+	public UnpackedBakedQuad bakeUnpacked()
+	{
+		UnpackedBakedQuad.Builder quad = new UnpackedBakedQuad.Builder( format.format );
+		pipe( quad );
+		return quad.build();
+	}
+
+	/**
+	 * A simple vertex format.
+	 */
+	public static class Vertex
+	{
+
+		public CachedFormat format;
+
+		/**
+		 * The raw data.
+		 */
+		public float[][] raw;
+
+		//References to the arrays inside raw.
+		public float[] vec;
+		public float[] normal;
+		public float[] color;
+		public float[] uv;
+		public float[] lightmap;
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Make lightmap a static final constant or non-public and provide accessors if needed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)